File: tooling/snapshot_toolchain.py
#!/usr/bin/env python3
"""Capture a toolchain snapshot into verification/toolchains/current.json.

The snapshot records the exact toolchain the repository's content was
verified against. It is generated, committed, and treated as the single
source of truth for version metadata everywhere else in the repository
(skill frontmatter, SKILL.md compatibility notes, README). Other tooling
checks consistency against this file; nothing else hand-writes versions.

Deterministic: output depends only on the installed toolchain, the host
platform, and the --date argument (no wall-clock reads without opt-in).

Usage:
  python3 tooling/snapshot_toolchain.py --date 2026-07-17 [--targets wasm-gc,wasm,js,native]
"""

from __future__ import annotations

import argparse
import json
import platform
import re
import subprocess
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent
SNAPSHOT_PATH = REPO_ROOT / "verification" / "toolchains" / "current.json"

# Two observed shapes:
#   "0.1.20260713 (75c7e1f 2026-07-13)"          (moon, moonrun; may carry a
#                                                 leading repeated tool name)
#   "v0.10.4+ade96c819 (2026-07-13)"             (moonc; commit after '+')
VERSION_RE = re.compile(
    r"^(?:[a-z-]+ )?v?(?P<version>[0-9][^ ]*) \((?:(?P<commit>[0-9a-f]+) )?(?P<date>\d{4}-\d{2}-\d{2})\)$"
)


def run(cmd: list[str]) -> str:
    return subprocess.run(
        cmd, check=True, capture_output=True, text=True
    ).stdout.strip()


def parse_component(item: dict) -> dict:
    raw = item["version"].strip()
    m = VERSION_RE.match(raw)
    if not m:
        raise SystemExit(f"unrecognized version string for {item['name']}: {raw!r}")
    return {
        "name": item["name"],
        "version": m.group("version"),
        "commit": m.group("commit") or "",
        "build_date": m.group("date"),
        "raw": raw,
    }


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--date", required=True, help="verification date, YYYY-MM-DD")
    parser.add_argument(
        "--targets",
        default="wasm-gc,wasm,js,native",
        help="comma-separated targets actually exercised during verification",
    )
    args = parser.parse_args()

    if not re.match(r"^\d{4}-\d{2}-\d{2}$", args.date):
        raise SystemExit("--date must be YYYY-MM-DD")

    version_all = json.loads(run(["moon", "version", "--all", "--json"]))
    components = [parse_component(item) for item in version_all["items"]]

    snapshot = {
        "_generated_by": "tooling/snapshot_toolchain.py (do not edit by hand)",
        "verification_date": args.date,
        "platform": {
            "os": platform.system(),
            "os_version": platform.mac_ver()[0] or platform.release(),
            "arch": platform.machine(),
        },
        "components": components,
        "verified_targets": sorted(args.targets.split(",")),
        "raw_version_all": run(["moon", "version", "--all"]),
    }

    SNAPSHOT_PATH.parent.mkdir(parents=True, exist_ok=True)
    SNAPSHOT_PATH.write_text(json.dumps(snapshot, indent=2, sort_keys=True) + "\n")
    print(f"wrote {SNAPSHOT_PATH.relative_to(REPO_ROOT)}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tooling/snapshot_toolchain.ts
import { mkdirSync, writeFileSync } from "node:fs";
import { dirname, join, relative } from "node:path";
import { exitWith, isMain, parseCliArgs, usageError } from "./lib/cli.ts";
import { stableJson } from "./lib/json.ts";
import { currentPlatform } from "./lib/platform.ts";
import type { CommandRunner } from "./lib/process.ts";
import { checkedOutput, runCommand } from "./lib/process.ts";
import { REPO_ROOT } from "./lib/repo.ts";

export const VERSION_RE =
  /^(?:[a-z-]+ )?v?(?<version>[0-9][^ ]*) \((?:(?<commit>[0-9a-f]+) )?(?<date>\d{4}-\d{2}-\d{2})\)$/;

interface VersionItem {
  name: string;
  version: string;
}

interface VersionOutput {
  items: VersionItem[];
}

export interface SnapshotComponent {
  name: string;
  version: string;
  commit: string;
  build_date: string;
  raw: string;
}

export function normalizeVersionPaths(raw: string, moonHome = process.env.MOON_HOME): string {
  if (moonHome === undefined || moonHome === "") {
    return raw;
  }
  return raw.replaceAll(moonHome.replace(/[\\/]+$/, ""), "<MOON_HOME>");
}

export function parseComponent(item: VersionItem): SnapshotComponent {
  const raw = item.version.trim();
  const match = VERSION_RE.exec(raw);
  if (match?.groups === undefined) {
    throw new Error(`unrecognized version string for ${item.name}: ${repr(raw)}`);
  }
  return {
    name: item.name,
    version: match.groups.version,
    commit: match.groups.commit ?? "",
    build_date: match.groups.date,
    raw,
  };
}

export function createSnapshot(
  date: string,
  targets: string[],
  runner: CommandRunner = runCommand,
): Record<string, unknown> {
  const versionJson = checkedOutput(runner, "moon", ["version", "--all", "--json"]);
  const versionOutput = JSON.parse(versionJson) as VersionOutput;
  const components = versionOutput.items.map(parseComponent);
  return {
    _generated_by: "tooling/snapshot_toolchain.ts (do not edit by hand)",
    verification_date: date,
    platform: currentPlatform(runner),
    components,
    verified_targets: [...targets].sort(),
    raw_version_all: normalizeVersionPaths(checkedOutput(runner, "moon", ["version", "--all"])),
  };
}

export function main(
  args = process.argv.slice(2),
  runner: CommandRunner = runCommand,
  repoRoot = REPO_ROOT,
): number {
  const usage =
    "usage: node tooling/snapshot_toolchain.ts --date YYYY-MM-DD [--targets TARGET,...]";
  const parsed = parseCliArgs(
    {
      args,
      options: {
        date: { type: "string" },
        targets: {
          type: "string",
          default: "wasm-gc,wasm,js,native",
        },
      },
      strict: true,
    },
    usage,
  );
  if (!parsed.ok) {
    return parsed.exitCode;
  }
  const { values } = parsed.result;
  if (values.date === undefined) {
    return usageError("--date is required", usage);
  }
  if (!/^\d{4}-\d{2}-\d{2}$/.test(values.date)) {
    return usageError("--date must be YYYY-MM-DD", usage);
  }

  const snapshotPath = join(repoRoot, "verification", "toolchains", "current.json");
  const snapshot = createSnapshot(values.date, values.targets.split(","), runner);
  mkdirSync(dirname(snapshotPath), { recursive: true });
  writeFileSync(snapshotPath, stableJson(snapshot), "utf8");
  console.log(`wrote ${relative(repoRoot, snapshotPath)}`);
  return 0;
}

function repr(value: string): string {
  return `'${value.replaceAll("\\", "\\\\").replaceAll("'", "\\'")}'`;
}

if (isMain(import.meta.url)) {
  exitWith(main());
}
